File: go/ff.go
package main

import (
  "fmt"
  "os"
  "path/filepath"
  "sync"
)

func searchFile(root, target string, results chan<- string, wg *sync.WaitGroup) {
  defer wg.Done()

  err := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
    if err != nil {
      return nil // Ignora erros de permissão
    }
    if d.IsDir() {
      // Para cada diretório, criamos uma nova goroutine para explorar ele em paralelo
      wg.Add(1)
      go searchFile(path, target, results, wg)
    } else if d.Name() == target {
      results <- path
    }
    return nil
  })

  if err != nil {
    fmt.Println("Erro ao percorrer:", err)
  }
}

func main() {
  if len(os.Args) < 3 {
    fmt.Println("Uso: programa <caminho> <arquivo>")
    os.Exit(1)
  }

  root := os.Args[1]   // Caminho base para busca
  target := os.Args[2] // Nome do arquivo a ser procurado

  results := make(chan string)
  var wg sync.WaitGroup

  // Inicia a busca com uma goroutine principal
  wg.Add(1)
  go searchFile(root, target, results, &wg)

  // Goroutine para fechar o canal quando todas as buscas terminarem
  go func() {
    wg.Wait()
    close(results)
  }()

  // Lê e imprime os resultados conforme são encontrados
  for path := range results {
    fmt.Println("Encontrado:", path)
  }
}

File: python/ff.py
import os
import sys
import multiprocessing

def search_file(root, target, queue):
    """Busca recursivamente pelo arquivo e coloca os resultados na fila."""
    try:
        for dirpath, _, filenames in os.walk(root, followlinks=False):
            if target in filenames:
                queue.put(os.path.join(dirpath, target))  # Encontra e adiciona à fila
    except PermissionError:
        pass  # Ignora diretórios sem permissão

def main():
    if len(sys.argv) < 3:
        print("Uso: python programa.py <caminho> <arquivo>")
        sys.exit(1)

    root = sys.argv[1]  # Diretório base
    target = sys.argv[2]  # Nome do arquivo
    queue = multiprocessing.Queue()  # Fila para armazenar os resultados

    # Criando múltiplos processos para acelerar a busca
    num_processes = multiprocessing.cpu_count()
    processes = []
    for _ in range(num_processes):
        p = multiprocessing.Process(target=search_file, args=(root, target, queue))
        p.start()
        processes.append(p)

    # Espera os processos terminarem
    for p in processes:
        p.join()

    # Coletando resultados
    results = []
    while not queue.empty():
        results.append(queue.get())

    # Exibe os resultados
    if results:
        print("\nArquivos encontrados:")
        for path in results:
            print(path)
    else:
        print("\nNenhum arquivo encontrado.")

if __name__ == "__main__":
    multiprocessing.set_start_method("fork", force=True)  # Evita bugs em Unix
    main()
